### ISSUE TITLE
Refactor and document facial flux infrastructure

### ISSUE THREAD

**Issue by MTCam** [illinois-ceesd/mirgecom#502]
Issue is resulting from discussions on #466.

We need a more unified and flexible infrastructure for handling "flux" calculations at element boundaries.  Users should be able to specify which sorts of numerical fluxes and quantity reconstructions are applied at element boundaries.  

Currently MIRGE-Com uses a local Lax-Friedrichs numerical flux for the inviscid fluxes used in the divergence operator at element boundaries.  Other valid choices could be Godunov, or Roe.  A central scheme is currently used for the divergence operator on the viscous fluxes, and the gradient operator.

**Comment by majosm** [comment]
Just adding a note here to think about later after talking about this a little with @inducer yesterday: it may be difficult to come up with a general interface for the flux functions, since each potentially needs different parameters (for example, LF needs `q` and `lambda`, whereas central doesn't). It might end up being easier to do something like pass a flag (i.e., `flux_type="lfr"`) instead?

**Comment by MTCam** [comment]
I think we should have the capability for users to pass their own flux reconstruction routines. If the generic `divergence_flux`, for example was like this:

```python
def divergence_flux(cv_tpair, flux_reconstruction):
	return flux_reconstruction(cv_tpair)@normal
```

Then the `flux_reconstruction` routine could be any of:
```python
# LFR routine just needs two fluxes, two states, and lambda
# This routine is fluid-specific and sets up the LFR computation
# with lam = max_wavespeed, as required.
def LFR_inviscid_flux_reconstruction(cv_tpair):
	flux_minus = inviscid_flux(cv_tpair.int)
    flux_plus = inviscid_flux(cv_tpair.ext)
	flux_tpair = TracePair(yadda blah blah)
	lam = max_local_wavespeed(cv_tpair)
	return lfr_flux(cv_tpair, flux_tpair, lam)

def central_inviscid_flux_reconstruction(cv_tpair):
	flux_minus = inviscid_flux(cv_tpair.int)
    flux_plus = inviscid_flux(cv_tpair.ext)
	flux_tpair = TracePair(yadda blah blah)
	return flux_tpair.avg

def user_inviscid_flux_reconstruction(cv_tpair):
	recon_cv = my_state_recon(cv_tpair)
	return inviscid_flux(recon_cv)
```

I've obviously left out some details, but I think the example captures the important bits.  Does something like this make sense?

**Comment by inducer** [comment]
Makes sense to me, and I'd be all in favor of it. The generic inputs to the numerical flux function are "my" and "opposite" values along with the normals, so a `TracePair` should work nicely.
